SI-2 Implement remaining health checks

sjoerdyoast opened the pull request:
## Context
<!--
What do we want to achieve with this PR? Why did we write this code?
-->

* This PR refactors the remaining legacy health checks to the new design from #17754 along with adding some small helper/factory classes to improve isolation in unit tests.

## Summary

<!--
Attach one of the following labels to the PR: `changelog: bugfix`, `changelog: enhancement`, `changelog: other`, `changelog: non-user-facing`.
If the changelog item is a bugfix, please use the following sentence structure: Fixes a bug where ... would ... (when ...).
If the changelog item is meant for the changelog of another add-on, start your changelog item with the name of that add-on's repo between square brackets, for example: * [wordpress-seo-premium] Fixes a bug where ....
If the changelog items is meant for the changelog of a javascript package, specify between square brackets in which package changelog the item should be included, for example: * [@yoast/components] Fixes a bug where ....
If the same changelog item is applicable to multiple changelogs/add-ons, add a separate changelog item for all of them.
-->
This PR can be summarized in the following changelog entry:

* Reïmplements the remaining legacy health checks to match the modern codebase.

## Relevant technical choices:

* The old health check implementations and interfaces are removed from the codebase.
* The health checks are no longer depending on the legacy `WPSEO_Options` class.
* The health checks now only use the dependency injection container for their dependencies.
* The `Ryte_Check` still depends on the legacy `WPSEO_Utils` class to preserve its old behavior.
* `WPSEO_Ryte` has been moved to `src/integrations/admin` as `Ryte_Integration`. The integration tests for it now also target the new class.
* The interface and implementation for `Ryte_Integration` hasn't changed apart from making `get_option()` public. The implementation for Ryte could definitely be improved but I consider that to be beyond the scope of this PR.

## Test instructions
<!--
Please follow these guidelines when creating test instructions:
- Please provide step-by-step instructions how to reproduce the issue, if applicable.
- Write step-by-step test instructions aimed at non-tech-savvy users, even if the PR is not user-facing.
-->


### Test instructions for the acceptance test before the PR gets merged
This PR can be acceptance tested by following these steps:

#### cURL - No premium plugins installed
- Make sure there are no premium plugins installed.
- Go to the Site Health page in the back-end.
- There shouldn't be a health check report regarding updating premium plugins.

#### cURL - No cURL version installed
- Make sure there are premium plugins installed.
- Make sure that cURL is not installed. To mimick this, go to `src/helpers/curl-helper.php` and change `is_installed()` to just return `false`.
- Go to the Site Health page in the back-end. There should be a failed health check saying that the Yoast premium plugins cannot update because the installed cURL version is probably too old.

#### cURL - No recent cURL version installed
- Make sure there are premium plugins installed.
- Make sure that the installed cURL version is lower than `7.34.0`. To mimick this, go to `src/helpers/curl-helper.php` and change `get_version()` to just return `'7.33.0'`.
- Go to the Site Health page in the back-end. There should be a failed health check saying that the Yoast premium plugins cannot update because the installed cURL version is probably too old.

#### cURL - Can't reach MyYoast API
- Make sure there are premium plugins installed.
- Make sure that the installed cURL version is more recent than 7.34.0.
- Make sure that the MyYoast API is unavailable. To mimick this, disconnect your computer from the internet.
- Go to the Site Health page in the back-end. There should be a failed health check saying that the Yoast premium plugins cannot update. It should also say that the installed cURL version is fine.

#### cURL - Health check passes
- Make sure there are premium plugins installed.
- Make sure that the installed cURL version is more recent than 7.34.0.
- Make sure that the MyYoast API is unavailable.
- Go to the Site Health page in the back-end. There should be a message saying that Yoast premium updates work fine.

#### Links table - Text link counter disabled
- Go to the back-end -> SEO -> General -> Features, and disable Text link counter.
- Go to the Site Health page. There shouldn't be a health check report regarding the Text link counter.

#### Links table - Text link counter enabled but not functional
- Make sure the Text link counter is enabled.
- Make sure that migrations haven't for the currently installed plugin version. To mimick this, go to `src/config/migration-status.php` and change `is_version()` to just return `false`.
- Go to the Site Health page. There should be a health check saying that the text link counter is not working because the plugin couldn't create the necessary tables automatically.

#### Links table - Health check passes
- Make sure the Text link counter is enabled.
- Go to the Site Health page. There should be a health check saying that the text link counter is working as expected.

#### Page comments - Page comments are not on a single page
- Go to the back-end -> Settings -> Discussion and enable "Break comments into pages..."
- Got to the Site Health page. There should be a health check saying that the comments break into multiple pages along with a recommendation to turn that off.

#### Page comments - Health check passes
- Make sure that "Break comments into pages..." is turned off.
- Go to the Site Health page. There should be a health check saying that comments are displayed on a single page.

#### Postname permalink - Permalinks don't contain the postname.
- Go to the back-end -> Settings -> Permalinks and set "Common Settings" to "Plain".
- Go to the Site Health page. There should be a health check saying that you don't have your postname in the URL of your posts and pages.

#### Postname permalink - Health check passes.
- Make sure that the "Common Settings" in Settings -> Permalinks is set to "Post name", or that it's set to "Custom Structure" and the structure contains `%postname%`.
- Go to the Site health page. There should be a health check saying that your permalink structure includes the post name.

#### Ryte - WordPress not in production mode
- Make sure that WordPress is in development mode. To mimick this, set the environment variable `WP_ENVIRONMENT_TYPE` to `development` (in the `docker-compose.yml` for example).
- Go to the Site Health page. There shouldn't be a health check about indexability.

#### Ryte - Yoast SEO is in development mode
- Make sure that WordPress in production mode (unset `WP_ENVIRONMENT_TYPE`).
- Make sure that Yoast SEO is in development mode
- Go to the Site Health page. There shouldn't be a health check about indexability.

#### Ryte - Website is not public
- Make sure that WordPress in production mode
- Make sure that Yoast SEO is in not in development mode
- Go to the back-end -> Settings -> Reading and enable "Discourage search engines from indexing this site".
- Go to the Site Health page. There shouldn't be a health check about indexability.

#### Ryte - Ryte is disabled
- Make sure that WordPress in production mode
- Make sure that Yoast SEO is in not in development mode
- Make sure that "Discourage search engines from indexing this site" is disabled.
- Go to the back-end -> SEO -> General -> Integrations and make sure that Ryte integration is set to "Off".
- Go to the Site Health page. There shouldn't be a health check about indexability.

#### Ryte - Error response from MyYoast
- Make sure that WordPress in production mode
- Make sure that Yoast SEO is in not in development mode
- Make sure that "Discourage search engines from indexing this site" is disabled.
- Make sure Ryte integration is enabled.
- Make sure that Ryte is unavailable. To mimick this, disconnect your computer from the internet.
- Go to the Site Health page. There should be a failing health check saying that an error occurred while checking whether your site can be found by search engines. The message should contain an error message. In the case of disconnecting from the internet, it should show a cURL error.

#### Ryte - Not indexable
- Make sure that WordPress in production mode
- Make sure that Yoast SEO is in not in development mode
- Make sure that "Discourage search engines from indexing this site" is disabled.
- Make sure Ryte integration is enabled.
- Make sure Ryte targets a non-indexable site. To mimick this, go to `src/admin/ryte/class-ryte-request.php` and go to the `get_remote()` function. Add the following line right before the return statement:
```php
$response = wp_remote_get( 'https://indexability.yoast.onpage.org/?url=https://yoast.com/dnxakdsgfsadg', $args );
```
- Go to the Site Health page. There should be a failing health check saying that your site cannot be found by search engines.

#### Ryte - Successful
- Make sure that WordPress in production mode
- Make sure that Yoast SEO is in not in development mode
- Make sure that "Discourage search engines from indexing this site" is disabled.
- Make sure Ryte integration is enabled.
- Make sure Ryte targets an indexable site. To mimick this, go to `src/admin/ryte/class-ryte-request.php` and go to the `get_remote()` function. Add the following line right before the return statement:
```php
$response = wp_remote_get( 'https://indexability.yoast.onpage.org/?url=https://yoast.com/', $args );
```
- Go to the Site Health page. There should be a passing health check saying that your site can be found by search engines.


### Test instructions for QA when the code is in the RC
<!--
Sometimes some steps from the test instructions for the acceptance test aren't relevant anymore once the code has been merged or the feature is complete. If that is the case, do not check the checkbox below.
QA is our Quality Assurance team. The RC is the release candidate zip that is tested before a release 
-->

* [x] QA should use the same steps as above.

<!--
If the above checkbox has not been checked, write down all steps QA should take to test this PR, not only the difference with the acceptance test steps. If QA should use the test instructions specified on the epic, paste a link to the relevant comment on the epic.
-->
QA can test this PR by following these steps:

*

## Impact check
<!--
Sometimes PRs have a bigger impact than is suggested in the user-facing changes. In such cases,
additional (regression) testing might be necessary. To make it clear what parts might need additional testing, please outline which parts of the plugin have been impacted by this PR.
-->
This PR affects the following parts of the plugin, which may require extra testing:

* `WPSEO_Ryte` has been moved to `src/integrations/admin` as `Ryte_Integration`. So explicit dependencies on `WPSEO_Ryte` are impacted. 
* No behavior of the health checks as they currently are has changed. The order in which they appear on the Site Health page may change, because it's now up to the dependency injection container to set up the health checks.

## UI changes

* [ ] This PR changes the UI in the plugin. I have added the 'UI change' label to this PR.

## Other environments

* [ ] This PR also affects Shopify. I have added a changelog entry starting with `[shopify-seo]` and I have added test instructions for Shopify.

## Documentation

* [ ] I have written documentation for this change.

## Quality assurance

* [x] I have tested this code to the best of my abilities
* [x] I have added unittests to verify the code works as intended
* [ ] If any part of the code is behind a feature flag, my test instructions also cover cases where the feature flag is switched off.

Fixes #

diedexx commented on src/services/health-check/curl-reports.php:
```diff
@@ -0,0 +1,113 @@
+	 * @param  WPSEO_Shortlinker      $shortlinker The WPSEO_Shortlinker object used to generate short links.
+	 * @return void
+	 */
+	public function __construct(
+		Report_Builder_Factory $report_builder_factory,
+		WPSEO_Shortlinker $shortlinker
+	) {
+		$this->report_builder = $report_builder_factory->create();
+		$this->shortlinker    = $shortlinker;
+	}
+
+	/**
+	 * Returns the message for a successful health check.
+	 *
+	 * @return string[] The message as a WordPress site status report.
+	 */
+	public function get_success_result() {
+		return $this->report_builder
+			/* translators: %1$s expands to 'Yoast'. */
+			->set_label( sprintf( __( '%1$s premium plugin updates work fine', 'wordpress-seo' ), 'Yoast' ) )
+			->set_status_good()
+			->set_description( __( 'Great! You can activate your premium plugin(s) and receive updates.', 'wordpress-seo' ) )
+			->build();
+	}
```
I think you can make this class less prone to side-effects by constructing a new builder every function call instead of reusing the same builder.
```suggestion
	public function __construct(
		Report_Builder_Factory $report_builder_factory,
		WPSEO_Shortlinker $shortlinker
	) {
		$this->report_builder_factory = $report_builder_factory;
		$this->shortlinker                     = $shortlinker;
	}

	/**
	 * Returns the message for a successful health check.
	 *
	 * @return string[] The message as a WordPress site status report.
	 */
	public function get_success_result() {
		return $this->report_builder_factory
			->create()
			/* translators: %1$s expands to 'Yoast'. */
			->set_label( sprintf( __( '%1$s premium plugin updates work fine', 'wordpress-seo' ), 'Yoast' ) )
			->set_status_good()
			->set_description( __( 'Great! You can activate your premium plugin(s) and receive updates.', 'wordpress-seo' ) )
			->build();
	}
```

sjoerdyoast replied:
That's true, but it will end up making the code a bit messier. Every report that the builder makes needs a `test_identifier`. Right now `set_test_identifier()` from Reports_Trait sets this on the builder instance.

I can omit that, and keep the `test_identifier` with the report class. Then I'd have to set the `test_identifier` whenever I create a new instance of a builder. If I then want to comply with Psalm's code analysis, I'd have to null check this value for every report as well. I could omit that by letting the report builder take `string|null` as an argument to its `set_test_identifier` method, but that seems a bit ugly to me.

So I understand the implications of reusing the same instance, but errors resulting from that are very unlikely to happen anyway, and making this change would make the code a bit less clean, especially when Psalm is involved.

I'm happy to hear your opinion on this though :)

diedexx replied:
You could require the identifier in the constructor of the reportbuilder and add it as a parameter for the factory as well. That way, it can't be null, right?

sjoerdyoast commented on config/dependency-injection/services.php:
```diff
@@ -30,6 +33,9 @@
 $container->register( WPSEO_Addon_Manager::class, WPSEO_Addon_Manager::class )->setFactory( [ Wrapper::class, 'get_addon_manager' ] )->setPublic( true );
 $container->register( WPSEO_Shortlinker::class, WPSEO_Shortlinker::class )->setFactory( [ Wrapper::class, 'get_shortlinker' ] )->setPublic( true );
+$container->register( WPSEO_Ryte::class, WPSEO_Ryte::class )->setFactory( [ Wrapper::class, 'get_ryte' ] )->setPublic( true );
+$container->register( WPSEO_Utils::class, WPSEO_Utils::class )->setFactory( [ Wrapper::class, 'get_utils' ] )->setPublic( true );
```
`is_development_mode` on the utils class checks if the plugin is in development mode. `Environment_Helper::is_production_mode` checks if WordPress is in production mode. They are two different things, but it's debatable whether it's useful to check both.

I'm already checking WordPress production mode with the statement `if (wp_get_environment_type() !== 'production')`, so I could replace that with the `is_production_mode()` method from the helper, and maybe omit the `is_development_mode()` method on the utils class. That way the health check would only care about the WordPress dev/production modes.

sjoerdyoast commented on src/services/health-check/ryte-runner.php:
```diff
@@ -0,0 +1,222 @@
+		WPSEO_Utils $utils
+	) {
+		$this->ryte                = $ryte;
+		$this->ryte_option_factory = $ryte_option_factory;
```
Moved `WPSEO_Ryte` to `src/integrations/admin` and made its `get_option()` method public for now. The entire Ryte integration could definitely use a refactor of its own, but we've considered it to be beyond the scope of this PR.

diedexx commented on wp-seo-main.php:
```diff
@@ -236,7 +239,7 @@ function _wpseo_activate() {
 
 	// Schedule cronjob when it doesn't exists on activation.
-	$wpseo_ryte = new WPSEO_Ryte();
+	$wpseo_ryte = new Ryte_Integration( new Options_Helper() );
```
You can use this to get the instance from the container instead
```suggestion
	$wpseo_ryte = YoastSEO()->classes->get( Ryte_Integration::class );
```

diedexx commented:
# Acceptance report

> cURL - No premium plugins installed
Make sure there are no premium plugins installed.
Go to the Site Health page in the back-end.
There shouldn't be a health check report regarding updating premium plugins.

This step fails: it is visible without a premium plugin if the check is successful. Which makes sense, looking at the code
https://github.com/Yoast/wordpress-seo/blob/47ef1e8bd35fee1c35a87d1f60c267ecb1107157/src/services/health-check/curl-check.php#L57-L63
Is this intended behaviour or should the order of the checks be inverted?

---


> Ryte - WordPress not in production mode
Make sure that WordPress is in development mode. To mimick this, set the environment variable WP_ENVIRONMENT_TYPE to development (in the docker-compose.yml for example).

This bit is in the middle of the test instructions without any expected behaviour. Could you please have a look where it should belong and what the expected behaviour is?
I'm assuming that there should be no "indexable" check with development mode enabled. I have tested this and the behaviour matches this assumption. Is my assumption correct?

For Ryte checks I've tested:
| \WPSEO_Utils::is_development_mode | WP_ENVIRONMENT_TYPE | extra| result |
| --- |---| ---|---|
| true | development | | no check |
| false | development | | no check |
| true | production | | no check |
| false | staging | | no check |
| false | production | Website not public | no check|
| false | production | Ryte responds with a 500 error if they can't resolve the domain name it seems. |"An error occurred while checking..."|
|false| production| With `$response = wp_remote_get( 'https://indexability.yoast.onpage.org/?url=https://yoast.com/dnxakdsgfsadg', $args );` | "Your site cannot be found by search engines" |
| false | production | With `$response = wp_remote_get( 'https://indexability.yoast.onpage.org/?url=https://yoast.com/', $args );` |Success:"Your site can be found by search engines"|

Did I miss anything?

sjoerdyoast commented:
> This step fails: it is visible without a premium plugin if the check is successful. Which makes sense, looking at the code

Good catch, I've fixed it now!

> This bit is in the middle of the test instructions without any expected behaviour. Could you please have a look where it should belong and what the expected behaviour is?

Your assumption is correct, I've updated the test instructions accordingly.

diedexx reviewed (approved):
Code changes look good 👍  Will start testing now

diedexx commented:
Acceptance OK 👍

diedexx commented:
The php 7.4 build is known to be broken on Travis right now. We've just merged a PR that switches to GH actions. I'll merge trunk in this branch to see that the build succeeds.
